File: akka/actor-system_2.4.8/src/main/java/demo/actor/life_cycle/ActorLifeCycleDemo.java
package demo.actor.scala.life_cycle;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.util.Timeout;
import scala.Option;
import scala.concurrent.duration.Duration;

import java.util.concurrent.TimeUnit;

class DemoActor extends UntypedActor {

    public void onReceive(Object message) throws Throwable {
        if (message instanceof String) {
            System.out.println("Message Received by " + getSelf().path().name() + " : " + message);
        }
        // code to generate exception. so that preRestart() and postRestart() will be invoked for self healing process.
        int i = 10 / 0;

    }

    public void preStart() throws Exception {
        super.preStart();
        System.out.println("preStart() is called! " + getSelf().path().name() + " is going to start.");
    }

    public void postStop() throws Exception {
        super.postStop();
        System.out.println("postStop() is called! " + getSelf().path().name() + " is stopped.");
    }

    public void preRestart(Throwable reason, Option<Object> message) throws Exception {
        super.preRestart(reason, message);
        System.out.println("preRestart() is called.");
        for (ActorRef ref : getContext().getChildren())
            getContext().stop(ref);
        postStop();
    }

    public void postRestart(Throwable reason) throws Exception {
        super.postRestart(reason);
        System.out.println("postRestart() is called.");
    }
}

/**
 * Driver Program for Actor Life Cycle Demo
 *
 * @author Kumar Gauraw
 */
public class ActorLifeCycleDemo {

    private static Timeout TIMEOUT = new Timeout(Duration.create(30, TimeUnit.SECONDS));

    public static void main(String[] args) throws Exception {

        // Create Actor System
        ActorSystem system = ActorSystem.create("ActorSystem");

        // Create Actor and Get Actor Reference
        ActorRef demoActor = system.actorOf(Props.create(DemoActor.class), "demoActor");

        // Send Message to the actor using tell()
        demoActor.tell("Hello Actor..", ActorRef.noSender());

        delay(10000);

        // Shutdown Actor System
        System.out.println("Shutting Down Actor System..");
        system.terminate();
    }

    /**
     * @param time
     */
    private static void delay(long time) {
        try {
            Thread.sleep(time);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: akka/actor-system_2.4.8/src/main/scala/demo/actor/scala/life_cycle/ActorLifeCycleDemo.scala
package demo.actor.scala.life_cycle

import akka.actor.{Actor, ActorRef, ActorSystem, Props}


class DemoActor extends Actor {
    def receive = {
        case msg: String => println("Message Received by " + self.path.name + " : " + msg)
            // code to generate exception. so that preRestart() and postRestart() will be invoked for self healing process.
            10 / 0;
    }

    override def preStart() {
        super.preStart();
        println("preStart method is called! " + self.path.name + " is going to start.");
    }

    override def postStop() {
        super.postStop();
        println("postStop method is called! " + self.path.name + " is stopped.");
    }

    override def preRestart(reason: Throwable, message: Option[Any]) {
        super.preRestart(reason, message);
        println("preRestart method is called!");
        println("Reason: " + reason);
    }

    override def postRestart(reason: Throwable) {
        super.postRestart(reason);
        println("postRestart is called");
        println("Reason: " + reason);
    }
}

/**
  * Driver Program for Actor Life Cycle Demo
  *
  * @author Kumar Gauraw
  */
object ActorLifeCycleDemo {
    def main(args: Array[String]): Unit = {
        val actorSystem: ActorSystem = ActorSystem("ActorSystem")
        val lifeCycleDemoActor: ActorRef = actorSystem.actorOf(Props[DemoActor], "lifeCycleDemoActor")
        lifeCycleDemoActor.tell("Hello", ActorRef.noSender)

        delay(10000)

        // Stop Actor
        println("Stopping Actor..")
        actorSystem.stop(lifeCycleDemoActor);

        // Shutdown Actor System
        println("Shutting Down Actor System..")
        actorSystem.terminate
    }

    private def delay(time: Long): Unit = {
        try {
            Thread.sleep(time)
        } catch {
            case e: Exception =>
                e.printStackTrace()
        }
    }
}


